presto-common/src/main/java/com/facebook/presto/common/PrestoVersion.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoVersion
+        implements Comparable<PrestoVersion>
+{
+    private static final Pattern VERSION_REGEX = Pattern.compile("^(0\\.([1-9][0-9]*))(\\.([1-9][0-9]*))?(\\.*)?(-([2][0-9]{7}\\.[0-2][0-9]{5}-[0-9]{1,}|SNAPSHOT).*)?$");
+    private int majorVersion;
+    private int minorVersion;
+    private String snapshotVersion;
+    private String version;
+
+    public PrestoVersion(String version)
+    {
+        Matcher matcher = VERSION_REGEX.matcher(version);
+        this.majorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(2) : null).orElse("0"));
+        this.minorVersion = Integer.parseInt(Optional.ofNullable(matcher.lookingAt() ? matcher.group(4) : null).orElse("0"));
+        this.snapshotVersion = Optional.ofNullable(matcher.lookingAt() ? matcher.group(7) : null).orElse("");
+        this.version = requireNonNull(version, "version string is null");
+    }
+
+    public String getVersion()
+    {
+        return version;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(this.majorVersion, this.minorVersion, this.snapshotVersion);
+    }
+
+    @Override
+    public boolean equals(Object version)
+    {
+        if (this == version) {
+            return true;
+        }
+        try {
+            PrestoVersion prestoVersion = (PrestoVersion) version;
+            return (Objects.equals(this.majorVersion, prestoVersion.majorVersion)
+                    && Objects.equals(this.minorVersion, prestoVersion.minorVersion)
+                    && (Objects.equals(this.snapshotVersion, prestoVersion.snapshotVersion)
+                    || Objects.equals(this.snapshotVersion, "SNAPSHOT")
+                        && !prestoVersion.snapshotVersion.isEmpty())
+                    || (Objects.equals(prestoVersion.snapshotVersion, "SNAPSHOT")
+                        && !this.snapshotVersion.isEmpty()));
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException("Please compare with a PrestoVersion object");
+        }
+    }
+
+    @Override
+    public int compareTo(PrestoVersion prestoVersion)
+    {
+        if (this.equals(prestoVersion)) {
+            return 0;
+        }
+        if (this.majorVersion != prestoVersion.majorVersion) {
+            return this.majorVersion < prestoVersion.majorVersion ? -1 : 1;
+        }
+        if (this.minorVersion != prestoVersion.minorVersion) {
+            return this.minorVersion < prestoVersion.minorVersion ? -1 : 1;
+        }
+        if (this.snapshotVersion != prestoVersion.snapshotVersion) {
```

Comment:
this is a reference check and not an equals check.
Unless the string is interned this will be always passing.
e.g. 	System.out.println(new String("abcd") != new String("abcd")); returns true.
You can use the `.equals` here